src/swap/useSwapQuote.ts

```diff
@@ -85,10 +170,81 @@ const useSwapQuote = () => {
           : null,
       }
 
-      return updatedExchangeRate
+      const baseApproveTx = createApproveTransaction(updatedField, quote.unvalidatedSwapTransaction)
+      const baseSwapTx = createSwapTransaction(quote.unvalidatedSwapTransaction)
+
+      let maxGasCosts: Array<[TokenBalance, BigNumber]> = []
+      for (const feeCurrency of feeCurrencies) {
+        // TODO: this is Celo specific, make it generic
+        const gasPrice = BigInt((await getGasPrice(feeCurrency.address ?? undefined)).toFixed())
+
+        const approveTx = {
+          ...baseApproveTx,
+          maxFeePerGas: gasPrice,
+          feeCurrency: !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined,
+          // We could estimate this ourselves,
+          // but for now we'll just use the value from the quote
+          gas: BigInt(quote.approveTransaction.gas),
+        }
+
+        const swapTx = {
+          ...baseSwapTx,
+          maxFeePerGas: gasPrice,
+          feeCurrency: !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined,
+        }
+
+        const maxGasCost = getMaxGasCost([approveTx, swapTx])
+        const maxGasCostInDecimal = new BigNumber(maxGasCost.toString()).shiftedBy(
+          -feeCurrency.decimals
+        )
+        maxGasCosts.push([feeCurrency, maxGasCostInDecimal])
+        if (maxGasCostInDecimal.gt(feeCurrency.balance)) {
+          // Not enough balance to pay for gas, try next fee currency
+          continue
+        }
+
+        const fromAmount = new BigNumber(approveTx.amountToApprove.toString()).shiftedBy(
+          -fromToken.decimals
+        )
+        if (
+          fromToken.tokenId === feeCurrency.tokenId &&
+          fromAmount.plus(maxGasCostInDecimal).gt(fromToken.balance)
+        ) {
+          // Not enough balance to pay for gas, try next fee currency
+          continue
+        }
+
+        // This is the one we can use
+        return {
+          ...baseQuoteResult,
+          type: 'possible',
+          approveTransaction: approveTx,
+          swapTransaction: swapTx,
+        } satisfies QuoteResultPossible
+      }
+
+      // So far not enough balance to pay for gas
+      // let's see if we can decrease the swap from amount
+      const result = maxGasCosts.find(([feeCurrency]) => feeCurrency.tokenId === fromToken.tokenId)
+      if (!result || result[1].gt(result[0].balance)) {
+        // Can't decrease the swap from amount
+        return {
+          ...baseQuoteResult,
+          type: 'not-enough-balance-for-gas',
+          feeCurrencies,
+        } satisfies QuoteResultNotEnoughBalanceForGas
+      }
+
+      // We can decrease the swap from amount to pay for gas,
+      // We'll ask the user if they want to proceed
+      return {
+        ...baseQuoteResult,
+        type: 'need-decrease-swap-amount-for-gas',
+        maxGasCost: result[1],
+        feeCurrency: result[0],
+      } satisfies QuoteResultNeedDecreaseSwapAmountForGas
```

Comment:
This is the main logic implemented.

Note: I haven't tried to break it down into smaller pieces. Although it's tailored for swaps, the logic should be able to be refactored to be more generic.